tic-tac-toe/src/urbit/app/toe.hoon

```diff
@@ -0,0 +1,1176 @@
+  ::
+  ?~  subs.sat
+    [~ this]
+  ::  the manual pull can only be sent by our current guest
+  ::
+  ?.  =(ze ze.i.subs.sat)
+    ::  if someone we didn't peer pokes us, do nothing
+    [~ this]
+  ::
+  =/  current-guest  (cite:title ze.i.subs.sat)
+  =^  edit  state.consol.sat  (transmit-sole reset)
+  ?~  t.subs.sat
+    ::  if we had only one opponent in the queue, reset
+    ::
+    =/  jon=(list [@t json])   [%status s+'null']~
+    =/  front-message  (send-tile-diff jon)
+    (wipe ~[(effect (all-effects edit)) front-message])
+  =/  new-guest  (cite:title ze.i.t.subs.sat)
+  =/  front-message=(list [@t json])
+    :~  [%message s+(crip new-guest)]
+        [%status s+'confirm']
+    ==
+  :_  %=  this
+        subs.sat  t.subs.sat
+        game.sat  %confirm
+      ==
+  :~  (send-tile-diff front-message)
+      %-  effect
+      :~  %mor
+          det+edit
+          klr+~[[[```%b] " [{current-guest} cancelled your game...] "]]
+          (prompt "{confirm}{new-guest} (Y/N)? | ")
+      ==
+  ==
+::
+::  %frontend
+::
+::    arms that deal with communication with eyre and the frontend
+::
++|  %frontend
+::
+++  bound
+  |=  [wir=wire success=? binding=binding:eyre]
+  ^-  (quip move _this)
+  [~ this]
+::
+::  $peer-toefile:
+::
+++  peer-toetile
+  |=  wir=wire
+  ^-  (quip move _this)
+  :_  this(front.sat ost.bol)
+  [ost.bol %diff %json *json]~
+::
+::  $poke-json:
+::
+++  poke-json
+  |=  jon=json
+  ^-  (quip move _this)
+  ?.  ?=(%o -.jon)
+    ::  ignores non-object json
+    [~ this]
+  =/  object=(map @t json)  +.jon
+  =/  data=json  (~(got by object) 'data')
+  =-  ~(action ge -)
+  ^-  (list @c)
+  ::  $data: cord->tape->(list @c)
+  ::
+  ::    this is a hack to reuse the parsing of console input
+  ::    to handle data validation. hacks need to be refactor
+  ::    at some point in the future, but not now
+  ::
+  ?+    -.data  !!
+    %a  =+  ((ar:dejs ni:dejs) data)
+        (tuba "{(scow %u (snag 0 -))}/{(scow %u (snag 1 -))}")
+    %s  (tuba (trip (so:dejs data)))
+  ==
+::
+::  +poke-handle-http-request: serve pages from file system based on URl path
+::
+::
+++  send-tile-diff
+  |=  pairs=(list [@t json])
+  ^-  move
+  =-  (snag 0 -)
+  %+  turn  (prey:pubsub:userlib /toetile bol)
+  |=  [=bone ^]
+  [bone %diff %json (pairs:enjs:format pairs)]
+::
+++  poke-handle-http-request
+  %-  (require-authorization:app ost.bol move this)
+  |=  =inbound-request:eyre
+  ^-  (quip move _this)
+  =/  request-line  (parse-request-line url.request.inbound-request)
+  =/  back-path  (flop site.request-line)
+  =/  name=@t
+    =/  back-path  (flop site.request-line)
+    ?~  back-path
+      ''
+    i.back-path
+  ::
+  :_  this  ^-  (list move)
+  ?~  back-path
+    [ost.bol %http-response not-found:app]~
+  ?:  =(name 'tile')
+    [ost.bol %http-response (js-response:app tile-js)]~
+  ?:  =(+<:site.request-line 'css')
+    [ost.bol %http-response (css-response:app style)]~
+  [ost.bol %http-response not-found:app]~
```

Comment:
Looks like you can preface this with
```hoon
:_  ~
:+  ost.bol  %http-response
```
for de-duplication.

Reply:
Sure; way cleaner! Also, I'll remove the CSS response; it's not used. (parts of my experimentation with not wanting to spend too much time in the frontend